lib/spack/spack/environment.py

```diff
@@ -1276,19 +1276,30 @@ def check_views(self):
     def _env_modifications_for_default_view(self, reverse=False):
         all_mods = spack.util.environment.EnvironmentModifications()
 
+        visited = set()
+
         errors = []
-        for _, spec in self.concretized_specs():
-            if spec in self.default_view and spec.package.installed:
-                try:
-                    mods = uenv.environment_modifications_for_spec(
-                        spec, self.default_view)
-                except Exception as e:
-                    msg = ("couldn't get environment settings for %s"
-                           % spec.format("{name}@{version} /{hash:7}"))
-                    errors.append((msg, str(e)))
-                    continue
-
-                all_mods.extend(mods.reversed() if reverse else mods)
+        for _, root_spec in self.concretized_specs():
+            if root_spec in self.default_view and root_spec.package.installed:
+                for spec in root_spec.traverse(deptype='run', root=True):
+                    if spec.name in visited:
+                        # It is expected that only one instance of the package
+                        # can be added to the environment - do not attempt to
+                        # add multiple.
+                        continue
```

Comment:
I wonder if we should warn or not if entering this branch. I can see arguments for both behaviors. so not sure what to do. I am fine deferring this choice to a later PR, leaving this comment in case you have a clear opinion on that.